velox/type/Variant.cpp

```diff
@@ -583,12 +584,43 @@ uint64_t variant::hash() const {
   if (a.isNull() or b.isNull()) {
     return false;
   }
+
+  // Use relative epsilon comparison.
+  // See
+  // https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/
+  // for details.
+
   if (a.kind_ == TypeKind::REAL) {
-    return fabs(a.value<TypeKind::REAL>() - b.value<TypeKind::REAL>()) <
-        kEpsilon;
+    float f1 = a.value<TypeKind::REAL>();
+    float f2 = b.value<TypeKind::REAL>();
+
+    // Check if the numbers are really close -- needed
+    // when comparing numbers near zero.
+    if (fabs(f1 - f2) < kEpsilon) {
+      return true;
+    }
+
+    float a1 = abs(f1);
+    float a2 = abs(f2);
+    float largest = a1 > a2 ? a1 : a2;
+
+    return fabs(f1 - f2) <= largest * FLT_EPSILON;
+  } else {
```

Comment:
shouldn't we test here if it's a double variant, and throw otherwise?

Reply:
This function is called only for REAL and DOUBLE:

```
  if ((kind_ == TypeKind::REAL) or (kind_ == TypeKind::DOUBLE)) {
    return equalsFloatingPointWithEpsilon(*this, other);
  }
```

Reply:
`equalsFloatingPointWithEpsilon()` is part of the public API though. What prevents someone from inadvertently calling it with a different type?

Reply:
It is now a free function in the anonymous namespace. I also added `VELOX_CHECK_EQ(a.kind(), TypeKind::DOUBLE);` to the 'else' branch there.